main/src/ca/uwaterloo/flix/api/lsp/provider/CodeActionProvider.scala

```diff
@@ -351,34 +348,74 @@ object CodeActionProvider {
   /**
     * Returns a code action to derive the `Eq` type class.
     */
-  private def mkDeriveEq(sym: Symbol.EnumSym, uri: String): CodeAction = mkDerive(sym, "Eq", uri)
+  private def mkDeriveEq(e: TypedAst.Enum, uri: String): Option[CodeAction] = mkDerive(e, "Eq", uri)
 
   /**
     * Returns a code action to derive the `Order` type class.
     */
-  private def mkDeriveOrder(sym: Symbol.EnumSym, uri: String): CodeAction = mkDerive(sym, "Order", uri)
+  private def mkDeriveOrder(e: TypedAst.Enum, uri: String): Option[CodeAction] = mkDerive(e, "Order", uri)
 
   /**
     * Returns a code action to derive the `ToString` type class.
     */
-  private def mkDeriveToString(sym: Symbol.EnumSym, uri: String): CodeAction = mkDerive(sym, "ToString", uri)
+  private def mkDeriveToString(e: TypedAst.Enum, uri: String): Option[CodeAction] = mkDerive(e, "ToString", uri)
 
   // TODO: Add derivation for the Hash and Sendable type classes.
 
   /**
-    * Returns a code action to derive the given type class `clazz` for the given enum symbol `sym`.
+    * Returns a code action to derive the given type class `clazz` for the given enum `e` if it isn't already.
+    * `None` otherwise.
     */
-  private def mkDerive(sym: Symbol.EnumSym, clazz: String, uri: String): CodeAction = CodeAction(
-    title = s"Derive $clazz",
-    kind = CodeActionKind.Refactor,
-    edit = Some(WorkspaceEdit(
-      Map(uri -> List(TextEdit(
-        Range(Position.fromEnd(sym.loc), Position.fromEnd(sym.loc)),
+  private def mkDerive(e: TypedAst.Enum, clazz: String, uri: String): Option[CodeAction] = {
+    val alreadyDerived = e.derives.classes.exists(d => d.clazz.name == clazz)
+    if (alreadyDerived) None
+    else Some(
+      CodeAction(
+        title = s"Derive $clazz",
+        kind = CodeActionKind.Refactor,
+        edit = Some(addDerivation(e, clazz, uri)),
+        command = None
+      )
+    )
+  }
+
+  /**
+    * Returns a workspace edit that properly derives the given type class, `clazz`, for the enum, `e`.
+    *
+    * For example, if we have:
+    * {{{
+    *   enum Abc {}
+    * }}}
+    * we could derive the type class 'Eq' like so:
+    * {{{
+    *   enum Abc with Eq { }
+    * }}}
+    *
+    * Or if there are already other derivations present:
+    * {{{
+    *   enum Abc with ToString {}
+    * }}}
+    * it will be appended at the end:
+    * {{{
+    *   enum Abc with ToString, Eq { }
+    * }}}
+    */
+  private def addDerivation(e: TypedAst.Enum, clazz: String, uri: String): WorkspaceEdit = {
```

Comment:
I'm looking at it right now, but I can't see a way to do it that wouldn't be unnecessarily complex, because it would need to be retrieved from the `root` or something, right? And then we'd have to match on the string anyway.

Reply:
We often just pass root around as an implicit parameter.

Reply:
Actually never mind. I take it back.

Reply:
OK, what we have now is good. So I think you are done.